File: pkg/api/stage_mapping.go
package orla

import "fmt"

// StageAssignment describes the backend and inference parameters assigned to a stage.
type StageAssignment struct {
	Backend         *LLMBackend
	MaxTokens       *int
	Temperature     *float64
	TopP            *float64
	ResponseFormat  *StructuredOutputRequest
}

// StageMappingInput is the input to a StageMapping: all stages that need assignment
// and the available infrastructure (backends).
type StageMappingInput struct {
	Stages   []*Stage
	Backends []*LLMBackend
}

// StageMappingOutput is the result of stage mapping: per-stage assignments.
type StageMappingOutput struct {
	Assignments map[string]*StageAssignment // stageID -> assignment
}

// StageMapping takes a set of stages and available backends, and assigns each stage
// to a specific backend with inference parameters. This is the planning step before
// workflow execution (Section 4.1 of the Orla paper).
type StageMapping interface {
	Map(input *StageMappingInput) (*StageMappingOutput, error)
}

// ExplicitStageMapping validates that every stage already has a backend assigned.
// This is the default mapping strategy when the user sets backends directly on stages.
type ExplicitStageMapping struct{}

// Map validates all stages have backends and returns their current assignments.
func (m *ExplicitStageMapping) Map(input *StageMappingInput) (*StageMappingOutput, error) {
	if input == nil {
		return nil, fmt.Errorf("stage mapping input cannot be nil")
	}
	out := &StageMappingOutput{
		Assignments: make(map[string]*StageAssignment, len(input.Stages)),
	}
	for _, s := range input.Stages {
		if s.Backend == nil {
			return nil, fmt.Errorf("stage %q (%s) has no backend assigned", s.Name, s.ID)
		}
		out.Assignments[s.ID] = &StageAssignment{
			Backend:        s.Backend,
			MaxTokens:      s.MaxTokens,
			Temperature:    s.Temperature,
			TopP:           s.TopP,
			ResponseFormat: s.ResponseFormat,
		}
	}
	return out, nil
}

// ApplyStageMappingOutput applies mapping output to stages, overwriting their backend
// and inference parameters with the assigned values.
func ApplyStageMappingOutput(stages []*Stage, output *StageMappingOutput) error {
	if output == nil {
		return fmt.Errorf("stage mapping output cannot be nil")
	}
	for _, s := range stages {
		assignment, ok := output.Assignments[s.ID]
		if !ok {
			continue
		}
		if assignment.Backend != nil {
			s.Backend = assignment.Backend
		}
		if assignment.MaxTokens != nil {
			s.MaxTokens = assignment.MaxTokens
		}
		if assignment.Temperature != nil {
			s.Temperature = assignment.Temperature
		}
		if assignment.TopP != nil {
			s.TopP = assignment.TopP
		}
		if assignment.ResponseFormat != nil {
			s.ResponseFormat = assignment.ResponseFormat
		}
	}
	return nil
}

File: pyorla/src/pyorla/stage_mapping.py
"""Stage mapping — pre-execution planning and backend assignment."""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from pyorla.types import LLMBackend, StructuredOutputRequest

if TYPE_CHECKING:
    from pyorla.stage import Stage


@dataclass
class StageAssignment:
    """Backend and inference parameters assigned to a stage."""

    backend: LLMBackend | None = None
    max_tokens: int | None = None
    temperature: float | None = None
    top_p: float | None = None
    response_format: StructuredOutputRequest | None = None


@dataclass
class StageMappingInput:
    """All stages that need assignment plus available backends."""

    stages: list[Stage] = field(default_factory=list)
    backends: list[LLMBackend] = field(default_factory=list)


@dataclass
class StageMappingOutput:
    """Per-stage assignments keyed by stage ID."""

    assignments: dict[str, StageAssignment] = field(default_factory=dict)


class StageMapping(ABC):
    """Base class for stage mapping strategies."""

    @abstractmethod
    def map(self, input: StageMappingInput) -> StageMappingOutput: ...


class ExplicitStageMapping(StageMapping):
    """Validates every stage already has a backend assigned."""

    def map(self, input: StageMappingInput) -> StageMappingOutput:
        output = StageMappingOutput()
        for stage in input.stages:
            if stage.backend is None:
                raise ValueError(
                    f"stage {stage.name!r} ({stage.id}) has no backend assigned"
                )
            output.assignments[stage.id] = StageAssignment(
                backend=stage.backend,
                max_tokens=stage.max_tokens,
                temperature=stage.temperature,
                top_p=stage.top_p,
                response_format=stage.response_format,
            )
        return output


def apply_stage_mapping_output(
    stages: list[Stage], output: StageMappingOutput
) -> None:
    """Apply mapping output to stages."""
    for stage in stages:
        assignment = output.assignments.get(stage.id)
        if assignment is None:
            continue
        if assignment.backend is not None:
            stage.backend = assignment.backend
        if assignment.max_tokens is not None:
            stage.max_tokens = assignment.max_tokens
        if assignment.temperature is not None:
            stage.temperature = assignment.temperature
        if assignment.top_p is not None:
            stage.top_p = assignment.top_p
        if assignment.response_format is not None:
            stage.response_format = assignment.response_format
